### PR TITLE
Adds BDX Protocol implementation

### PR DESCRIPTION
**Pull request by Apollon77**
This PR mainly adds the BDX protocol implementation. We support most of the features from the specification with the following notes:
* Even that Matter allows 64bit values for size and offset, we support up to MAX_SAFE_INTEGER for size and offset (which basically is 2^53 - 1 and so far enough for us).
* We support partial transfers (startOffset or shorter dataLength) only when we act as the sender. As a receiver,
  only full transfers are supported - the main reason was that the Streams in JavaScript can not easiely "seek" to other positions in the file, this only works with a FileSystemStream in Node.js (as example). Also because I doubt that really anyone uses or expect this we decline such transfers (as allowed by spec) nd can implement whenever needed
* We do not use BlockQueryWithSkip when requesting data ourselves (See above)

Additionally the PR also optimizes an edge case in the DataReadQueue and ensures that StandaloneAcks are always only considered for Secure Channel protocol.

BDX also got extensive tests.

it is also notable that this is currently not yet used anywhere, so it is the pure functionality.

The use cases for later steps are:
* Devices trigger transfers from OTA provider cluster. Device does ReceiveInit to OTA provider (server) with the URL it got from a QueryImage command
* Logs upload: Device does a SendInit to the node that requested the logs (RetrieveLogsRequest on Diagnostic Logs cluster with protocol of BDX)

So maybe more adjustments (like Filedesignator-mapping) or such needs to be added later then

### CODE REVIEW

File: packages/protocol/src/bdx/schema/BdxAcceptMessagesShema.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
+/**
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/protocol/src/bdx/schema/BdxAcceptMessagesShema.ts:1]: valid, will be fixed

File: packages/protocol/src/bdx/FileDesignator.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        return this.#fd;
+    }
+
+    /** Return the textual/string representation of the FileDesignator. */
+    get text(): string {
+        const fileDesignatorData = Bytes.of(this.#fd);
+        // When all uint8 values are in char() range "a-z0-0-." then use this as the blob name, else hex encode it
+        const isValidName = fileDesignatorData.every(
+            byte =>
+                (byte >= 0x11 && byte <= 0x5a) || // A-Z
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/protocol/src/bdx/FileDesignator.ts]: valid, will be fixed

File: packages/protocol/src/bdx/BdxMessenger.ts [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        const acceptMessage = await this.nextMessage([BdxMessageTypes.ReceiveAccept]);
+        const result = BdxReceiveAcceptMessage.decode(acceptMessage.payload);
+        logger.debug("Received Bdx ReceiveAccept", result);
+        return result;
+    }
+
+    /** Encodes and sends a Bdx SendAccept message. */
+    async sendSendAccept(message: BdxSendAccept) {
+        logger.debug("Sending Bdx SendAccept", message);
+        await this.exchange.send(BdxMessageTypes.SendAccept, BdxSendAcceptMessage.encode(message));
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxMessenger.ts]: I might generate these methods but I like the clean explicit layout.  It does feel like you could create a message type <-> schema mapping and have encode/decode handled automatically...

Encode seems easy as it's just lookup and no types necessary.  And integrated with in a send wrapper it could be invisible

Decode -- not here but elsewhere -- would be slightly more complicated but you could parse automatically and with a big discriminated union you could get types for free after checking message type

Buut fine as future TODO or as is, certainly easy to follow

File: packages/protocol/src/bdx/BdxMessenger.ts [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    get maxPayloadSize() {
+        return this.#exchange.maxPayloadSize;
+    }
+
+    /**
+     * Waits for the next message and returns it.
+     * A List of allowed expected message types can be provided.
+     * If the message type is not in the list, an error will be thrown.
+     */
+    async nextMessage(expectedMessageTypes: number[], timeout = this.#messageTimeout, expectedMessageInfo?: string) {
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxMessenger.ts]: Why just number here?

File: packages/protocol/src/bdx/BdxMessenger.ts [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        const message = await this.exchange.nextMessage({ timeout });
+        const messageType = message.payloadHeader.messageType;
+        if (expectedMessageTypes !== undefined && expectedMessageInfo === undefined) {
+            expectedMessageInfo = expectedMessageTypes.map(t => `${t} (${BdxMessageTypes[t]})`).join(",");
+        }
+        this.throwIfErrorStatusReport(message, expectedMessageInfo);
+        if (expectedMessageTypes !== undefined && !expectedMessageTypes.includes(messageType))
+            throw new UnexpectedDataError(
+                `Received unexpected message type: ${messageType}, expected: ${expectedMessageInfo}`,
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxMessenger.ts]: Could include the textual name here, e.g `${BdxMessageType[messageType] ?? "unknown"}#${messageType}` or similar...  Maybe in a helper function so it's formatted consistently everywhere

File: packages/protocol/src/bdx/BdxMessenger.ts [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        logger.debug("Waiting for Bdx Block");
+        const message = await this.nextMessage([BdxMessageTypes.Block, BdxMessageTypes.BlockEof]);
+        const messageType = message.payloadHeader.messageType;
+        const block = BdxBlockMessage.decode(message.payload);
+        if (messageType !== BdxMessageTypes.BlockEof && block.data.byteLength === 0) {
+            // empty block only allowed in BlockAckEof
+            throw new BdxError("Received empty data in Block message", BdxStatusCode.BadMessageContent);
+        }
+        logger.debug(`Received Bdx Block with ${block.data.byteLength} bytes`, block);
+        return { ...block, messageType };
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxMessenger.ts]: This is where a discriminated union would be handy, so caller could switch on messageType to get the proper type without cast.

File: packages/protocol/src/bdx/BdxProtocol.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    async onNewExchange(exchange: MessageExchange, message: Message) {
+        const initMessageType = message.payloadHeader.messageType;
+        const messenger = new BdxMessenger(exchange);
+
+        switch (initMessageType) {
+            case BdxMessageTypes.SendInit:
+            case BdxMessageTypes.ReceiveInit:
+                logger.debug(
+                    `Initialize Session for ${BdxMessageTypes[initMessageType]} message on BDX protocol for exchange ${exchange.id}`,
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxProtocol.ts]: Should BdxMessageTypes be singular?

File: packages/protocol/src/bdx/BdxProtocol.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+                try {
+                    await bdxSession.processTransfer();
+                } catch (error) {
+                    logger.error(`Error processing BDX transfer:`, error);
+                }
+
+                break;
+            default:
+                logger.warn(`BDX session for exchange ${exchange.id} already exists, sending error`);
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxProtocol.ts]: It's not clear to me how the default case in this switch statement gets to this message

File: packages/protocol/src/bdx/BdxMessenger.ts [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+    close() {
+        return this.#exchange.close();
+    }
+
+    /**
+     * Ensure that the value is a number and that it is not too large. Matter spec allows also 64bit values, but they
+     * make little sense for now, so make sure we handle them as too large. MAX_SAFE_INTEGER is 2^53-1 and is
+     * enough for now.
+     */
+    asNumber(value: number | bigint | undefined, context = "", bdxErrorCode = BdxStatusCode.Unknown): number {
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxMessenger.ts]: Maybe call validatedSize or something related to purpose of number?

File: packages/protocol/src/bdx/BdxMessenger.ts [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+        if (typeof value !== "number" && typeof value !== "bigint") {
+            throw new InternalError(`${context} ${value} is not a number`); // Should not happen
+        }
+        if (value > Number.MAX_SAFE_INTEGER) {
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxMessenger.ts:286]: If we could enforce maxTransferSize here we could avoid redundant checks/error message in other places

File: packages/protocol/src/bdx/FileDesignator.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+                (byte >= 0x41 && byte <= 0x5a) || // A-Z
+                (byte >= 0x61 && byte <= 0x7a) || // a-z
+                (byte >= 0x30 && byte <= 0x39) || // 0..9
+                byte === 0x2e || // "."
+                byte === 0x2d, // "-"
+        );
+        if (isValidName) {
+            return fileDesignatorData.reduce((name, byte) => name + String.fromCharCode(byte), "");
+        } else {
+            return Bytes.toHex(fileDesignatorData);
```
Comment (lauckhart) [inline packages/protocol/src/bdx/FileDesignator.ts]: Maybe need a protected prefix on names so e.g. "\x12" and "12" don't map to same file?

File: packages/protocol/test/bdx/bdx-mock-exchange.ts [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * A mock message exchange for the BDX protocol.
+ *
+ * This provides an intermediate level of mock communication, higher level than network mocking but lower level than
+ * direct method calls.
+ *
+ * Our mocking of the message exchange context is a bit half assed, we don't use a real channel or session and just
+ * stub out methods as necessary.
+ */
+export class MockExchange extends MessageExchange {
```
Comment (lauckhart) [inline packages/protocol/test/bdx/bdx-mock-exchange.ts:42]: Is this similar enough to the other mock exchange that we could make it part of package like MockNetwork and have it be shared?

File: packages/protocol/src/bdx/BdxSession.ts [hunk 13]
```diff
@@ -0,0 +1,575 @@
...
+    #messenger: BdxMessenger;
+    #isSender: boolean;
+    #isInitiator: boolean;
+    #initMessage?: BdxInit;
+    #started = false;
+    #closed = AsyncObservable();
+    #isClosed = false;
+
+    /** Configuration for transfer behavior */
+    #bdxConfig: BdxSession.Config = {
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxSession.ts]: Exporting as a constant near the Options interface would surface defaults a little more clearly

File: packages/protocol/src/bdx/BdxSession.ts [hunk 14]
```diff
@@ -0,0 +1,575 @@
...
+ * Notes:
+ * * Even though Matter allows 64bit values for size and offset, we do not use them, as they make no sense for now.
+ *   We support up to MAX_SAFE_INTEGER for size and offset (which basically is 2^53 - 1 and so far enough for us).
+ * * We support partial transfers (startOffset or shorter dataLength) only when we act as the sender. As a receiver,
+ *   only full transfers are supported.
+ * * If data are queried with skipping some data (BlockQueryWithSkip) then we might send out a message shorter than the
+ *   blockSize which is ok according to spec.
+ * * We do not use BlockQueryWithSkip when requesting data ourselves
+ */
+export class BdxSession {
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxSession.ts:34]: This is a large class...  It's nice with the flows extracted...  A few potential other components stand out for extraction:

- BdxSessionConfiguration - covers the stuff in the constructor and the type definitions int he BdxSession namespace
- BdxSessionInitiator - covers the bulk of the code in private methods here to do setup, before invoking `this.#processTransferFlow()` (which would then only be invoked in one place)
- Could also decouple persistence, although that's perhaps more for separation of concerns than for readability

File: packages/protocol/src/bdx/flow/BdxTransferFlow.ts [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+ * @license
+ * Copyright 2022-2025 Matter.js Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { BdxError, BdxMessenger } from "#bdx/index.js";
+import { BdxStatusCode } from "#types";
+
+/** Base class for BDX transfer flows. */
+export abstract class BdxTransferFlow {
```
Comment (lauckhart) [inline packages/protocol/src/bdx/flow/BdxTransferFlow.ts]: Cool abstraction.  "Transfer" seems a little redundant with "Flow", maybe could just call "BdxFlow"?  Mainly because it'd make the subclass names less noisy.  I'd leave off Bdx too given directory layout but that feels more like personal preference.

File: packages/protocol/src/bdx/flow/BdxDrivenSendingTransferFlow.ts [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+            if (this.isClosed) {
+                return; // Session is closed, so we stop processing
+            }
+
+            const blockCounter = this.nextMessageCounter;
+            // Read the next data chunk from the storage
+            ({ data, done, bytesLeft } = await this.readDataChunk(reader, blockSize, bytesLeft, dataLength));
+
+            if (done) {
+                // Send the last Block and wait for AckEof and close down
```
Comment (lauckhart) [inline packages/protocol/src/bdx/flow/BdxDrivenSendingTransferFlow.ts]: I'd just have the break here, put other stuff after the loop, or see my comment in BdxDrivingReceivingTransferFlow.ts for slightly higher-level organization

File: packages/protocol/src/bdx/flow/BdxDrivingReceivingTransferFlow.ts [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+import { BdxWritingTransferFlow } from "./BdxWritingTransferFlow.js";
+
+/**
+ * BDX Transport flow logic for a "BDX Driving Receiver":
+ * - Requests the next Blow using BlockQuery and then reads the Block
+ * - Last message is confirmed with an AckEof
+ */
+export class BdxDrivingReceivingTransferFlow extends BdxWritingTransferFlow {
+    async processTransfer() {
```
Comment (lauckhart) [inline packages/protocol/src/bdx/flow/BdxDrivingReceivingTransferFlow.ts]: An alternate design for these flows could be methods in the base class:

```ts
// Stuff in processTransfer() before while (true) updates local vars here
abstract initTransfer(): void;

// Stuff inside while loop goes here (except check for isClosed goes in base class); returns boolean indicating whether loop should break
abstract transferNextChunk(): Promise<boolean>;

// Stuff in if (done) goes here
abstract finalizeTransfer(): Promise<void>;
```

...thinking being that you have the loop and check for isClosed() in the base class and then it's easy to see the transfer stages in each flow implementation.

File: packages/protocol/src/bdx/flow/BdxFollowingSendingTransferFlow.ts [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+import { BdxWritingTransferFlow } from "./BdxWritingTransferFlow.js";
+
+/**
+ * BDX Transport flow logic for a "BDX Following Sender":
+ * - Reads the Block and sends an Ack (if synchronous transfer is used)
+ * - Last message is confirmed with an AckEof
+ */
+export class BdxFollowingSendingTransferFlow extends BdxWritingTransferFlow {
```
Comment (lauckhart) [inline packages/protocol/src/bdx/flow/BdxFollowingSendingTransferFlow.ts]: I was a little confused by "reading" vs "writing", maybe could call base classes "inbound" and "outbound" instead?

Part of the my confusion I think comes from a transfer flow that is "followingsending" being named from the perspective of the peer, while "writingtransferflow" is named from the perspective of the server.  But "inbound" to me makes it clear data moves from the peer.

Unless I'm still confused. ;)

File: packages/protocol/src/bdx/flow/BdxReadingTransferFlow.ts [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+            // Done needs to be true when value is undefined or there is something broken
+            if (!done) {
+                throw new BdxError(
+                    `Data length too short, expected ${blockSize}bytes, but got less`,
+                    BdxStatusCode.LengthTooShort,
+                );
+            }
+            value = new Uint8Array(); // Simulate an empty value when we reached the end of the stream
+        } else if (value.byteLength < blockSize) {
+            // When we git less data than blocksize it is the last block, so validate that
```
Comment (lauckhart) [inline packages/protocol/src/bdx/flow/BdxReadingTransferFlow.ts]: "get" :)

File: packages/protocol/src/protocol/MessageExchange.ts [hunk 20]
```diff
@@ -88,6 +88,9 @@ export type ExchangeSendOptions = {
 
     /** Additional context information for logging to be included at the beginning of the Message log. */
     logContext?: ExchangeLogContext;
+
+    /** Allows to override the protocol ID of the message, mainly used for standalone acks. */
+    protocolId?: number;
```
Comment (lauckhart) [inline packages/protocol/src/protocol/MessageExchange.ts:95]: It's weird to me that the message exchange is protocol independent except for this one feature.  Would it be possible to create a `SecureProtocolMessageExchange` that adds standalone ack support?  Or to do it with composition, have some sort of plugin component with hooks in all of the places where it's wired in?

I know standalone ack is pretty well intertwined with the code here so maybe better as a future TODO.

**Review by lauckhart — APPROVED** [review]

File: packages/protocol/src/bdx/BdxMessenger.ts (hunk 3 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/BdxMessenger.ts]: I tried around  a lot but without changing a lot of things - including how Schema works and such - is was not really able to come up with anything better than https://github.com/matter-js/matter.js/pull/2369/commits/ddc7a1fb9153411fbfc25d44852051fb6c5d7583 ... Yes it streamlines that class a bit more but in fact for the cost of several "as unknown as Type" things because in fact especially when reading the multiple options are meehh.

In fact logic and code wise it is still ok because all the single methods act kind of as type guard so that internally such type castings are no risk ... WDYT?

File: packages/protocol/src/bdx/BdxMessenger.ts (hunk 4 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/BdxMessenger.ts]: valid, fixed

File: packages/protocol/src/bdx/BdxProtocol.ts (hunk 7 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/BdxProtocol.ts]: makes sense

File: packages/protocol/src/bdx/BdxProtocol.ts (hunk 8 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/BdxProtocol.ts]: fixed

File: packages/protocol/src/bdx/BdxMessenger.ts (hunk 9 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/BdxMessenger.ts]: asSafeNumber ?!

File: packages/protocol/src/bdx/BdxMessenger.ts (hunk 10 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/BdxMessenger.ts:286]: The method is not only used for maxTransferSize and mainly used to get the bigint case out of the way. Or what did you mean? There are also not that many checks for maxTransferSize

File: packages/protocol/src/bdx/FileDesignator.ts (hunk 11 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/FileDesignator.ts]: valid

Comment (Apollon77) [reply packages/protocol/src/bdx/FileDesignator.ts]: "__hex__" ?? or do you think "0x" as prefix is enough

File: packages/protocol/src/bdx/flow/BdxTransferFlow.ts (hunk 15 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/flow/BdxTransferFlow.ts]: killed both. just left Bdx prefix in one place for logger

File: packages/protocol/src/bdx/flow/BdxDrivingReceivingTransferFlow.ts (hunk 17 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/flow/BdxDrivingReceivingTransferFlow.ts]: That worked good :-) Thanks

File: packages/protocol/src/bdx/flow/BdxFollowingSendingTransferFlow.ts (hunk 18 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/flow/BdxFollowingSendingTransferFlow.ts]: Yeeahh I was confused too but because of the structure of the BDX protocol with the "Driver" vs "Follower" and "Sender" vs "Receiver" things get mixed. Also Inbound/Outbound is not really helping, because still leave interpretation

My Idea with "Reading" and "Writing" was to more exact say what "we do" ... are we "Reading" data from the storage or are we "writing" data into the Storage.

File: packages/protocol/src/protocol/MessageExchange.ts (hunk 20 above)
Comment (Apollon77) [reply packages/protocol/src/protocol/MessageExchange.ts:95]: Yeahh I tink it is because the Standalone Ack is part of the MRP feature of the "underlaying protocol scheme". So SecureProtocol is the underlaying scheme. BDX is more like an "Application protocol" which sits on top of a Secure Channel layer.
This means that even if you process BDX messages the underlaying MRP layer belongs to SecureChannel.
I will add a TODO. In fact we might need to get the "protocol binding" out of the exchange and more formalize the messages we send because in fact the messages then are "protocol bound" and not the exchange.

File: packages/protocol/src/bdx/BdxSession.ts (hunk 14 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/BdxSession.ts:34]: Lets see what you say to this :-) Large refactoring incoming

File: packages/protocol/src/bdx/FileDesignator.ts (hunk 11 above)
Comment (lauckhart) [reply packages/protocol/src/bdx/FileDesignator.ts]: 0x seems fine

File: packages/protocol/src/bdx/schema/BdxMessage.ts [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+        kind: T,
+    ): asserts message is BdxMessage<T> {
+        if (message.kind !== kind) {
+            throw new InternalError(
+                `Expected BDX message of type ${BdxMessageType[kind]}, but got ${BdxMessageType[message.kind]}(${message.kind})`,
+            );
+        }
+    }
+
+    export const Kinds: { [key: number]: Schema<any, Bytes> } = {
```
Comment (lauckhart) [inline packages/protocol/src/bdx/schema/BdxMessage.ts]: Could maybe make "any" and "Bytes" be default generic parameters?

File: packages/protocol/src/bdx/BdxMessenger.ts [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+    close() {
+        return this.#exchange.close();
+    }
+
+    /**
+     * Ensure that the value is a safe number and that it is not too large. Matter spec allows also 64bit values, but they
+     * make little sense for now, so make sure we handle them as too large. MAX_SAFE_INTEGER is 2^53-1 and is
+     * enough for now.
+     */
+    asSafeNumber(value: number | bigint | undefined, context = "", bdxErrorCode = BdxStatusCode.Unknown): number {
```
Comment (lauckhart) [inline packages/protocol/src/bdx/BdxMessenger.ts]: Maybe `asSafeInteger` for naming as there is no notion of "safe" otherwise.  Or could be "asByteLength" or something as that's what it's used for.

I would also make it static as I don't think there's any references to local fields.  I think this confused me before because I assumed it was used locally

File: packages/protocol/src/bdx/BdxSession.ts (hunk 14 above)
Comment (lauckhart) [reply packages/protocol/src/bdx/BdxSession.ts:34]: I like it

**Review by lauckhart — APPROVED** [review]

File: packages/protocol/src/bdx/schema/BdxMessage.ts (hunk 21 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/schema/BdxMessage.ts]: I pin the Bytes

File: packages/protocol/src/bdx/BdxMessenger.ts (hunk 22 above)
Comment (Apollon77) [reply packages/protocol/src/bdx/BdxMessenger.ts]: As discussed in Discord we leave as is
